### ISSUE TITLE
State of the API (April 11 2016)

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#123]
_This ticket assesses the current state of the API, and the path forward from here._
## Basic Goal

The fundamental question here is "how can we get the next version out as quickly as possible?" In that question the "next version" is defined by the lowest common denominator that will allow us to begin building modules for integrating with other plugins in a forward-composible manner. It doesn't have to include any shiny gadgets or whistles, but it needs to be possible for us to add such new features and have all of our modules dance to whatever tune core plays for them.
## Problems

The difficulty in such an information-driven composible framework is knowing what information we will need ahead of time, without knowing how we will be needing to recompose the modular components in the future. The meta-information that will be in WordPoints's core will only be able to describe the behavior of those components in relation to the information that they supply. 

However, building all of the features now would take a huge amount of time, and at the same time partly undermines the need for forward-composibility (though of course the feature set will certainly grow and evolve no matter how much we offer from the start).

If it was possible for us to provide all of the information that could possibly exist about a particular modular component, we could simply do that, and be certain that our meta-information would always have the terms available to describe the information provided by the modules. However, that isn't possible, because the information which can exist about a modular component is partly determined by the declarations of the meta-information itself, since it can define entirely new concepts to which the modular components may be applied but cannot comprehend in advance. (Related to [subject-oriented programming](https://en.wikipedia.org/wiki/Subject-oriented_programming).)

Due to the mutual and relational nature of this sort of modulatiry then, it is impossible to achieve perfect forward-recomposibility.
## Proposed Solutions

However, we can do two things. First, we can, as we have been doing, attempt to think ahead as much as possible in terms of the information that we provide. And second, we can provide a structural framework in which the information and meta-information (and logic) can exist, regardless of the particular nature of any of it. That way, when we do inevitably need to add information and meta-information (and logic) in the future, we will have the best possible chance of doing so in a manner that is fully backward compatible, and thus will be able to utilize the existing information without special extension.

Ideally, all of the information would  be defined by the plugins themselves, but of course WordPress doesn't provide a framework for that. This means that it would be possible for our information about he plugin to become desynchronized with the plugin itself. That is a problem that may in some cases be unavoidable. However, in other cases it may be possible to extract this information from the plugin directly, at a low level. In many cases of course this may be impossible, or inviable due to potential fragility.

**Comment by JDGrimes** [comment]
## Basic Concepts

At a conceptual level, plugins of the sort which we are speaking are made up of entities or "objects", and different actions or events that relate to those objects. These objects then define or exist within certain contexts. They can also relate to one another. The nature of the entity, exactly what it is, is defined by its "attributes". 

It would seem then that entities and actions do indeed make for the basic concepts within any framework for our purposes.

**Comment by JDGrimes** [comment]
## Composibility

But that part of the structure is just the conceptual part. We also need composibility in how we arrange entity features, etc., as well as information-driven architecture. This transcends entities and actions or anything else we would build.

I think the first step is developing a compositional framework. Currently we can compose different whole objects to build a feature, but we really would like to also compose different units to build an object. I suppose that that is where traits might come in handy. However, they are not composed at run-time, which may be a useful feature for us in our framework. (Related is [this SO answer which shows how traits could be used to support dependency injection](http://stackoverflow.com/a/7893384/1924128) of a particular feature across classes without having to duplicate code.)

**Comment by JDGrimes** [comment]
## Information Composition

Over time, we're going to need to add more information about entities. At present the entity storage API is an example. The only way to add this information will be either by the consumer deducing it via a low-level API or else by an update of the code which defines those entities and/or provides their information. Our goal is to provide a framework for the entity information that will be as extensible as possible.

In this, things to consider are the amount of memory that information will take up, and also the different ways in which it will be used. If we store different information about entities separately, it will allow each set of information to be loaded only when it is needed. Overall this is probably a minor concern, however as the amount of information an entity may need to provide about itself grows, it seems unnecessary to have all of the information in memory, much of which may be unneeded for the feature that is currently in operation. Also, composition of this sort might even allow us to better organize our inheritance tree, since different information sets for an entity may be better described via differential inheritance.

However, placing the entity data into separate classes increases the number of objects that we create, and the number of classes (and thus files) that we load. And I'm unsure that we would gain anything at all from it, because we'd either have to have a separate class for each entity and data type or else load the data into the class from an array. But if we were to do the former, we'd end up with a huge number of files to load, and if we did the latter, we'd have most of the data in memory anyway. Unless we store the data elsewhere. However, if we do that, we either have to have a bunch of files, as in the case above, or else we have to store all of the data together, which means that it would all load unnecessarily when we only needed the info for a single entity.

This is the biggest issue with an information-driven API: how to store the information.

It is true that at present memory issues really don't present a problem for us, so we could just ignore it and only come back to it later if it ever really does present a problem, which may not be likely. There is also the option of not really storing the information at all, but deducing it at some low level at run-time. The problem with that is that it may not be possible, and that it would take longer. We could introduce some sort of caching, sure, but then that defeats the purpose as far as memory. But perhaps there is some greater purpose here than memory conservation—like the ability to automatically integrate with a plugin that we've never heard of before, at run-time. The problem, as noted above, is that it really isn't possible (at least, it isn't practical without AI) to get a list of entities that can exist within the site at run-time. So we have to know at least some information about the entities to start with. It is likely possible, however, to determine some of the information that we want using just the most basic information about each entity, though not all of the info we want could be obtained that way. The benefit of doing that wherever we can though is greater forward-compatibility as well as auto-compatibility of a feature with every existing entity when that feature and the new information that it needs about entities is introduced.

The most basic information about entities, from which we would be able to derive some information about them, would be where they are stored. Given just a table name for instance, we could determine what attributes that entity has, what context it existed in, etc. The thing that would stop us from doing some of that would be that we cannot translate entity names out of thin air, nor entity attribute names. Some information about them could be mined, but we do have to supply that along with other basic stuff. Still, it is something to consider when introducing new features that require new entity information in the future. Though I think that mainly the information which would be needed there would be in relation to how an entity was used, not so much what it was.

**Comment by JDGrimes** [comment]
Storing raw information in specifically named classes seems ludicrous though. I think that we should consider abstracting information out from the classes themselves, and then let the plugin determine how to handle that information, that is, what classes contain the logic to pair up with that information, at run-time. So basically we make a composite object which can handle the logic, based on the information.
However, it isn't just information that entity classes provide, they also generally provide some utility as well. And we also run into i18n again. These are the reasons that we ended up with the current structure to start with, and at the moment I'm not sure that there is really much that we can do to improve it.

I think that in order to know whether what we have here is sufficient, we'd first need to know exactly how we intend to use this information. Which means that we need to get the other aspects of the API as we want them, and then see how things are working. So in part we need to wait until other information is introduced that doesn't have to be stored on the entity classes before we can really develop that kind of framework, or at least before we can apply it to the entities. Meaning that this may not need to be part of the initial version.

Whatever we do, it would probably be useful to be able to change how the information is stored in the future.  So we should split implementation from the needs for the information. That once again forces us to answer the question as to what we need from the information, of course.

**Comment by JDGrimes** [comment]
## Entity Change API vs Action API

Our entity change API is also important, and although entities are usually stored in the database, the entity change API can be triggered listening for database queries, by the WordPress action API, by a REST API (for foreign entities), or by any mechanism really.

In that sense it is a part of our ideal action API, one which can listen for an event via any possible mechanism. However, it only listens for actions/events that modify entities, it doesn't listen for events that just relate to entities in some way. That's where an action API is still needed.

Our current implementation of the action API is designed to handle both of these things, but with a heavy action flavor. It is sort of intended to be more a meta API able to handle an entity change action API as well as a regular event API. But at the same time it is specifically designed with WordPress actions in mind. I think this is something that has to be resolved. The problem is that we may need to keep it somewhat like that in order to get the initial version of this out ASAP. But on the other hand, I fear that the result would be an API that would handle neither actions nor entity changes well. That may be acceptable however, if we get the information for plugin integration available in the process, so that it could be used via whatever API we design in the future. The important thing is to know as best we can what info we need and to provide a framework for the future. One part of that is knowing what entities exist and providing a basic framework for working with them, the other part is knowing what actions occur, and providing a basic framework for working with them.

Currently, there are two issues with the hooks API:
1. It is too specific to the WordPress action and filter API. Certainly that should be a main focus, but a more generally interface with broader application would be nice.
2. It is too generic in the way that it forces entity changes to use the same API as events. There may be room for some meta API here but I think that right now both of these are suffering.

**Comment by JDGrimes** [comment]
## Supplying data through methods

It seems odd that classes that primarily offer information do so only through a bunch of methods. Since they are information containers, they would be much easier to deal with if they offered this information publicly via properties. Entities are an example of this.

The only potential issue of course is that public properties can be set by outside code, which isn't a feature that we specifically want. However, it is possible that this will actually be a welcome side-effect. The main issue I see is that it is more error prone, since it is possible to modify the data unintentionally. If we are concerned about that though, perhaps we could just block this via defining `__set()` and `__unset()`. Then we could still define setters if we wanted to allow outside forces to modify this information. But since it is part of the class definition, I think that it is generally not a good idea to allow modification at all, especially when we don't actually keep the same instance in memory throughout the execution of the script.

But magic methods add overhead, and note that we'd have to define `__get()` as well because the properties would have to be protected or else `__set()` wouldn't apply to them at all. But given the added overhead, this really isn't much different that having getter methods to start with.

Just making the properties public can also make it easier for us to export the data to the UI, etc. However, I'm not sure that we actually need to export all of the data. But still, as far as the parts we need, it could be easier.

This means that the value for the property cannot be supplied dynamically though. Of course, I suppose that usually it could be determined and set at the time that the class was constructed, but this may be wasteful in circumstances when the value of the property isn't even needed.

Other than that, I'd say that we might want to consider going ahead and doing this, implementing the magic methods. Even though it will be more overhead at present, there is always the hope that PHP will implement some kind of read-only properties in the future. And in that case, I guess we would be prepared.

In https://github.com/WordPoints/hooks-api/issues/124#issuecomment-214513449 we considered the possibility of using class constants instead of public properties. The problem was that it wouldn't be compatible with PHP 5.2, at least not in a way that would allow for the kind of object composition that we are proposing.

The benefit of constants over properties though was that they could be defined in interfaces. This is one reason that using public properties is not as good. It breaks down the class/interface structure, since interfaces can't defined properties.

**Comment by JDGrimes** [comment]
> Currently, there are two issues with the hooks API:
> 1. It is too specific to the WordPress action and filter API. Certainly that should be a main focus, but a more generally interface with broader application would be nice.
> 2. It is too generic in the way that it forces entity changes to use the same API as events. There may be room for some meta API here but I think that right now both of these are suffering.

Perhaps then, the router should trigger the event API, but not encompass it in any way. Right now it does. So splitting that out might be a first step.

But before we go too far, we're going to have to make a decision as to how far we're going to go—how much of the API are we going to change, and how much of it is going to stay the same? If we really want to split this all the way, then we might be forced to introduce the full-fledged entity change API. Though perhaps we can come up with a creative way to continue routing that through the event API temporarily.

For example, what about action types? Do we keep them or not? Again, this depends upon whether we're keeping a partly unified API and UI. The UI is subject to the split too, although I suppose that it wouldn't have to be split even if the underlying API was, in theory.

Our real focus here though is no longer so much the UI or API at all, but rather the underlying information in the entity and events. That's what we want to get settled as quickly as possible, so that we can start releasing modules. Then we can transform the API however we want. (Though that isn't entirely without its own set of compatibility concerns, of course.)

**Comment by JDGrimes** [comment]
## Event arg type API

This last thought is why we decided to focus on the entity storage API, which is now completed (or at least in a quasi-stable state for now).

As far as an action/event API goes, I think that we'd probably benefit from a similar counterpart to this there as well. Perhaps there should be some means of specifying not just what entities were involved in the action, but perhaps _how_ they were involved. Why? Because in the future it would allow us to separate contextual bystanders from entities performing an action from an entity that is being modified during the action. Actually, I guess actions that involve entity modification would generally be listened for using the entity change API. But there is still the need to listen for the event itself, that is, the event of the user clicking the publish button, if we want to listen for this. But the point is that there are different ways that entities relate to each event. They can just be context, like site, network, etc., or they can be the thing performing the event, like a user, or the thing experiencing the event or that the event is performed relative to. Anyway perhaps this is something to explore further in a separate ticket.

**Comment by JDGrimes** [comment]
## The cell analogy

Here is an analogy I came up with while I was thinking about information and composition:

Think of the plugin as a cell, and other plugins as other cells. The site can be an organ or the organism as a whole (thinking of it as an organ only for multisite, but maybe that would be better thought of as a colony of organisms).

Like a cell, our plugin needs to collect information about its environment. This information will then be handled via internal information and meta-information to produce a response (or indifference) in the cell. We can listen to changes in the environment at the database level, or communication form other cells  like actions. An action is just a change in the information state of the environment.

So we basically have
- info
- meta-info
- logic

**Comment by JDGrimes** [comment]
I think this ticket has mainly run its course. The only thing that we'd ideally like to do but may not pursue in the initial version is providing a better generic framework for information providers and consumers. Possibly we'll do that in a later version of the API, when we have some more real-world experience with how exactly this will all work. It is pretty hard for us to do anything meaningful at this point without much in the way of real use-cases. We just have to use our imaginations.
